File: python/198.打家劫舍.py
#
# @lc app=leetcode.cn id=198 lang=python3
#
# [198] 打家劫舍
#

# @lc code=start
class Solution:
    # 一维数组，如果追求空间效率，还可以用滚动数组
    def rob(self, nums: List[int]) -> int:
        n = len(nums)
        if n < 2:
            return nums[0]
        ans = [0] * n
        ans[0] = nums[0]
        ans[1] = max(ans[0], nums[1])
        for i in range(2, n):
            ans[i] = max(ans[i - 2] + nums[i], ans[i - 1])
        return ans[n - 1]

    # 二维数组
    def rob1(self, nums: List[int]) -> int:
        n = len(nums)
        ans = [[0, 0] for _ in range(n)]
        ans[0][0] = 0
        ans[0][1] = nums[0]
        for i in range(1, n):
            ans[i][0] = max(ans[i - 1][0], ans[i - 1][1])
            ans[i][1] = ans[i - 1][0] + nums[i]
        return max(ans[n - 1][0], ans[n - 1][1])
# @lc code=end

File: cpp/198.打家劫舍.cpp
/*
 * @lc app=leetcode.cn id=198 lang=cpp
 *
 * [198] 打家劫舍
 */
#include <algorithm>
#include <iostream>
#include <string>
#include <unordered_map>
#include <unordered_set>
#include <vector>

using namespace std;
// @lc code=start
class Solution
{
public:
    int rob(vector<int>& nums)
    {
        if (nums.size() == 1) {
            return nums[0];
        }

        int pre     = max(nums[0], nums[1]);
        int pre_pre = nums[0];
        int ans = pre;

        for (int i = 2; i < nums.size(); i++) {
            ans = max(pre_pre + nums[i], pre);
            pre_pre = pre;
            pre = ans;
        }

        return ans;
    }

    /**
     * 自写
    */
    int rob2(vector<int>& nums) {
        if (nums.size() == 1) {
            return nums[0];
        }

        int steal_max     = nums[0];
        int pre_steal_max = nums[0];

        int not_steal_max     = 0;
        int pre_not_steal_max = 0;

        for (int i = 1; i < nums.size(); i++) {
            steal_max     = max(pre_steal_max, pre_not_steal_max + nums[i]);
            not_steal_max = max(pre_steal_max, pre_not_steal_max);

            pre_steal_max     = steal_max;
            pre_not_steal_max = not_steal_max;
        }

        return max(steal_max, not_steal_max);
    }
};
// @lc code=end
